### PR TITLE
Fix put/get functions for non-contiguous Mat on iOS/macOS

### PR DESCRIPTION
**Pull request by komakai**
### Pull Request Readiness Checklist

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.

Basically the same fix as #19503 for iOS and macOS

### CODE REVIEW

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:
